### ISSUE TITLE
Account configuration and validation

### ISSUE THREAD

**Issue by laurb9** [prebid/prebid-server#1395]
## Account Configuration

We should have account configuration settings available in Prebid Server.

In addition to validating that an account (publisher id) is known to the prebid-server, account-specific configurations would be needed to support planned functionality. The Java version has the following per-account configurations, for example:

- eventsEnabled: to control win notification events such as #1202 or video impression tracking #1015
- enforceGdpr
- bannerCacheTtl
- videoCacheTtl
- analyticsSamplingFactor

PBS-Go has a simple yaml accounts list in legacy code, not used for openrtb2 endpoints and with the current definition as a list of strings, only useful for validation. PBS-Java supports a yaml file with a more complex structure, and a database source.

### Implementation

For implementation, supporting the same data sources as PBS-Java could be useful to some extent and the ability to refresh the data in flight is needed as well. The HTTP api source is well suited for this purpose, even more versatile than the database in my opinion. FileFetcher less so (I will create a ticket for separate discussion on this).

I'm willing to take a stab at implementing this. Thoughts ?

**Comment by bretg** [comment]
Thanks @laurb9 - I imagine @SyntaxNode would happily take the help here.

However, expanding the schema presents its own problems - another enhancement under way in PBS-Java is to make the account query configurable.

Unlike the Stored Request queries, the Account query in PBS-Java (and I think PBS-Go too) is currently hard-coded. This isn't ideal because:

1) it causes upgrade problems between versions
2) doesn't allow host companies to define their own schemas as easily
3) it's inconsistent with the other query definitions

So what we're building:
- a new configuration parameter will be added "settings.database.account-query" for specifying SQL query that loads account data. No default value, i.e. host company must provide it in its deployment.
- certain restrictions are imposed on the SQL query and described in documentation (order, number and data types of returned attributes, placeholder for account ID in a WHERE clause, a LIMIT clause etc.)
- there will be one macro in the query which is %ACCOUNTID%

e.g.
```
settings.database.account-query =SELECT uuid, price_granularity, banner_cache_ttl, video_cache_ttl, events_enabled, enforce_ccpa, tcf_config, analytics_sampling_factor, truncate_target_attr FROM accounts_account where uuid = %ACCOUNTID% LIMIT 1
```

**Comment by SyntaxNode** [comment]
I completely agree. The simple account list was "good enough" for a while, but now that we're expanding per-account settings for privacy (GDPR, CCPA, etc..) we need a much improved solution.

Your timing is impeccable. We've recently begun building a new account service, similar to the approach taken by PBS-Java. That work is being spearheaded by @bsardo. Would be interested in working with him on this?

The scope of our work is currently limited to a configurable account service, GDPR privacy settings, and a static file fetcher. I think it would be fantastic if you could provide either a db or http fetcher (or both) and implement the white list functionality for the OpenRTB endpoints.

@bsardo will open a draft PR soon with the framework he's been working on and for you to review and collaborate.

How does that sound?

**Comment by laurb9** [comment]
@SyntaxNode sounds great! I started work on the file fetcher refactor, as it looked easy enough to add accounts in the same fashion as stored reqs, but I'm also interested in the http account config.

@bretg I am not going to use db configuration but I can take a look at the code - I still think a default query should be in the config to guide admins, otherwise it would be difficult for people to figure out what is expected. Also, have you considered a json object type as db schema ? postgres supports json (raw) / jsonb (parsed) so then the data format can be defined only once.

**Comment by laurb9** [comment]
@SyntaxNode @bsardo I've done a POC in this draft https://github.com/prebid/prebid-server/pull/1426, is this what you had in mind ?

It runs but there is more work to be done for sure, and I would like to know if I'm going in the right direction. I am using the existing eager fetcher so we can focus on the account bits.

The idea is to load account configs from separate json (or yaml) files/db/api sources, like stored requests, using the fetcher framework, and caching them in a new cache. 
This can replace both the list of valid accounts and the blacklist, the latter can be done simply by creating account structures with disabled=false, either from config builder or as individual files.

**Comment by laurb9** [comment]
The PR is now ready #1426 :

- Defines the Account object with per-publisher settings
- Adds a cfg.DefaultAccount for global settings that are overrideable per publisher
- Expands account validation to require a valid account not just any non-empty value
- Makes the resolved account object available to auction and analytics
- Cache TTL can now be controlled at account level (this was mostly so I have a reason to pass account to HoldAuction, as it is a one-line change otherwise)
- Uses fetcher framework to retrieve accounts from individual <id>.json files or HTTP API
  (database to be added later)

Example configuration for accounts stored in `stored_requests/data/by_id/accounts/<id>.json`
```yaml
 accounts:
   filesystem:
     enabled: true
     directorypath: "./stored_requests/data/by_id"
```

Example account that would go in `stored_requests/data/by_id/accounts/valid_acct.json`
```json
{
    "ID": "valid_acct",
    "name": "test account",
    "disabled": false,
    "ttl": {
        "banner": 600,
        "video": 3600,
        "native": 3600,
        "audio": 3600
    },
    "events": {
        "enabled": true
    }
}
```

**Comment by SyntaxNode** [comment]
Keeping this open until we implement:

- [ ] GDPR account level controls.
- [ ] CCPA account level controls.
- [ ] Event account level controls. (In Progress)
- [ ] analyticsSamplingFactor account level controls.

**Comment by bsardo** [comment]
@laurb9 To prevent duplicate effort, I just wanted to point out that I'm beginning work on GDPR and CCPA account level controls.

**Comment by SyntaxNode** [comment]
Closing. PBS-Go has the same account configuration values as PBS-Java for implemented features. We're tracking the account config for features which have yet to be implemented as part of their feature story. If you notice a difference, please open a new issue.
